examples/nidaqmx_analog_input/measurement.py

```diff
@@ -17,7 +17,7 @@
 service_directory = pathlib.Path(script_or_exe).resolve().parent
 measurement_service = nims.MeasurementService(
     service_config_path=service_directory / "NIDAQmxAnalogInput.serviceconfig",
-    version="0.1.0.0",
+    version="1.0.0.0",
```

Comment:
You can delete this parameter entirely from the `measurement.py` files. @nick-beer made it optional in https://github.com/ni/measurement-plugin-python/pull/827 . If not specified, it will be read from the `.serviceconfig` file.

Reply:
Yeah - in fact, passing the a version to the constructor is problematic _if_ it is not also found in the .serviceconfig file.  We should probably update the service.py code to enforce this.  The problem is as follows:

- You specify a version in your constructor, but _not_ in your .serviceconfig file.
- When the service is published to plugin library or statically installed, the DiscoveryService thinks the service is _unversioned_ - because there's no version in the .serviceconfig file.
- As a result, it registers it internally with a default version (used to be `1.0.0`, but as of last night, it's now `0.0.1`)
- However, when the service starts itself and registers itself with discovery service, it uses this version that's passed in the constructor, and _those versions don't match_.  This means DiscoveryService fails to see this as a valid registration/resolution, and the client is left hanging, waiting for that valid registration.

Reply:
We previously discussed emitting a warning if you specify the version when constructing MeasurementService: https://github.com/ni/measurement-plugin-python/pull/827#discussion_r1704562044

Maybe that would be appropriate now, but this seems like a regression for existing services.

Reply:
> @nick-beer made it optional in https://github.com/ni/measurement-plugin-python/pull/827

The version parameter isn't optional yet. It would be a breaking change to take it out, wouldn't it?
```
Traceback (most recent call last):
  File "C:\dev\measurementlink-python\examples\game_of_life\measurement.py", line 16, in <module>
    measurement_service = nims.MeasurementService(
TypeError: __init__() missing 1 required positional argument: 'version'
```

Reply:
Pretty sure it is optional - see [here](https://github.com/ni/measurement-plugin-python/blob/82ab5530e68e40e1b86ca6ff4e20761f2c4d5788/packages/service/ni_measurement_plugin_sdk_service/measurement/service.py#L180).

```python
class MeasurementService:
    """Class that supports registering and hosting a python function as a gRPC service."""

    def __init__(
        self,
        service_config_path: Path,
        version: str = "",
        ui_file_paths: List[Path] = [],
        service_class: Optional[str] = None,
    ) -> None:
```

Reply:
It's optional in the 2.1 pre-release:
```python
class MeasurementService:
    """Class that supports registering and hosting a python function as a gRPC service."""

    def __init__(
        self,
        service_config_path: Path,
        version: str = "",
        ui_file_paths: List[Path] = [],
        service_class: Optional[str] = None,
    ) -> None:
```
But game_of_life is still using the released 2.0 package, not the 2.1 pre-release.

Reply:
> Yeah - it is a bit of a regression. I think an appropriate fix is to allow the following and raise an error for anything else:
> 
> * No version specified in either the constructor or serviceconfig.  This will just get the default version
> * No version specified in the constructor, but a version specified in the serviceconfig.
> * A version specified in the constructor, and that same version is specified in the serviceconfig
> 
> If someone has an existing service with a version in the constructor (effectively all services), they'll be required to make some kind of change when updating SDK versions. This is unfortunate, but because DiscoveryService reads the .serviceconfig file to get version information, I'm not sure how else to get around it.

The current state of this PR should fix the examples. Updating the error logic in a follow-up PR is fine with me.

Reply:
https://github.com/ni/measurement-plugin-python/issues/908